stellargraph/layer/graph_classification.py

```diff
@@ -250,3 +243,123 @@ def in_out_tensors(self):
         return x_inp, x_out
 
     build = deprecated_model_function(in_out_tensors, "build")
+
+
+@experimental(reason="Missing unit tests and generally untested.")
+class DeepGraphConvolutionalNeuralNetwork(GCNSupervisedGraphClassification):
```

Comment:
I did consider this option and I decided against it for several reasons. 

First of all, I don't like monstrous model classes that try to implement many different models in one by considering different combinations of parameter values. For example, if in the future we want a model that replaces the GCN layers with GAT or another layer, then we have to add more parameters. I find that having to keep track that the correct combination of parameters is given for each model implemented by a class is going to be difficult to maintain in the future and difficult to explain to the end-user. 

In addition, model classes should be named after the algorithm implemented; so just like we have model classes for GAT, GCN, GraphSAGE, APPNP, etc. the DGCNN model has a clearly identifiable name in the literature (although I did opt out for using the full name instead of the acronym) and our model class should be named after it instead of hiding it behind the generic `GCNSupervisedGraphClassification` name. 

Finally, it was no accident that I could implement DGCNN by subclassing the `GCNSupervisedGraphClassification` class; it was by design and I prefer to keep it this way.

Reply:
It's a good point that it's good to have names that match the papers for ease of reference/reproduction.
 
However, I'm concerned that we're going to end up with an explosion of models for each paper as special cases each with their own implementation, rather than distilling out the core pieces that slot together in reusable ways (especially when the pieces are so orthogonal, in this case). For instance, other models use outputs from multiple convolution layers (e.g. UGraphEmb) suggesting it's a useful/reuseable technique.

I think we can work together find a middle ground that offers the benefits of both:

- the flexible model class, to reduce how much work we do, and offer expert users full control
- named/easy-to-access models for published papers

For instance, we could have the `GCNSupervisedGraphClassification` class support all of the arguments (with sensible defaults), and then subclasses that slot in the appropriate arguments in the `__init__`, and then don't need to override any other functions:

```python
class DeepGraphConvolutionalNeuralNetwork(GCNSupervisedGraphClassification):
    """
    Args:
        ...
        k (int): size (number of rows) of output tensor.
        ...
        kwargs: any other arguments for :class:`GCNSupervisedGraphClassification`.
    """
    def __init__(self, layer_sizes, activations, k, generator, **kwargs):
        super().__init__(
            layer_sizes=layer_sizes,
            activations=activations,
            generator=generator,
            pooling=SortPooling(k=k),
            pool_all_gcn_layers=True,
            **kwargs,
       )
```

Instead of a subclass, this could even be a function:

```python
def deep_graph_convolutional_neural_network(layer_sizes, activations, k, generator, **kwargs):
    return GCNSupervisedGraphClassification
        layer_sizes=layer_sizes,
        activations=activations,
        generator=generator,
        pooling=SortPooling(k=k),
        pool_all_gcn_layers=True,
        **kwargs,
    )
```

What do you think?